docs/history-service.md

```diff
@@ -0,0 +1,283 @@
+```
+
+However, there are several requirements that make the above challenging to implement. In particular:
+
+- **Scaling**: the History service must be able to support arbitrarily many concurrent workflow executions. See [History shards](#history-shards).
+- **Consistency**: the Workflow and Activity tasks in Matching Service must be consistent with the workflow execution data in History Service. See [Consistency guarantees](#consistency-guarantees).
+- **Latency** of request handling and server-side workflow advance must be minimized.
+
+The rest of this document describes the implementation, which addresses these requirements.
+
+The History Service has many other responsibilities which are not covered here, for example: Shard Membership and Ownership, Replication and conflict resolution, Visibility, Archival, Workflow Reset.
+
+## Concepts
+
+### History shards
+
+A single Temporal Cluster manages many (e.g. millions of) Workflow Executions.
+These are logically partitioned into subsets that we call History Shards.
+The Temporal Cluster typically has multiple instances of the History service application process (each typically running on a separate machine/VM), and each History service instance owns some set of History Shards.
+"Owning" a History Shard means being responsible for the lifecycle of every workflow execution in the shard by [handling incoming requests](#rpc-handling) relating to that workflow execution, managing timers, and causing Workflow and Activity Tasks to be enqueued in Matching Service. To achieve this, each shard manages its own [task queues](#queue-processing).
+
+Therefore, objects corresponding to History Shards exist in the memory of a History service instance.
+Persistence details are abstracted away via a Go interface but in practice, when the persistence layer is backed by Cassandra, History Shards map 1-1 with partitions of the [`executions` table](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/schema/cassandra/temporal/schema.cql#L52). Persistence layer implementations also exist for MySQL, Postgres, and SQLite.
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- [`schema/`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/schema) directory
+- When the History service starts on a host, [`Start()`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/history_engine.go#L288) is called for each shard owned by that host.
+
+</details>
+
+### Workflow Execution "Mutable State"
+
+For every workflow execution, we maintain a collection of data structures summarizing various aspects of its current state, for example, the identities of in-progress activities, timers, and child workflows.
+Although most of this data could in principle be recomputed from Workflow History Events when handling an incoming request, this would be slow, and hence the summaries themselves are persisted.
```

Comment:
I think the actual reason we do this is because of Cassandra. We have some relational data, like the pending activites for a workflow, and we want to do 2 relational DB-esque things that Cassandra doesn't really support:
1. Update this relational data (e.g. add a pending activity) transactionally
2. Query this data in an indexed fashion (e.g. get all child workflows of a workflow)

In a normal RDBMS, you'd store this data in separate tables, with a logical data model (e.g. an activities table with a foreign key on the workflow). However, these things don't really exist, and there's no working analog in Cassandra:
1. You can't update data in different tables in a single transaction in Cassandra, so we keep almost everything in one polymorphic "executions" table
2. Even if we didn't need transactional updates, we wouldn't be able to query this data efficiently because adding secondary indices in Cassandra has so many drawbacks, that we just almost never do it. Instead, we configure it so that all data you need is queryable via <full partition key> + <some prefix of the clustering key>, or it's colocated in the same row.

So, you're right that we keep track of some state, which could be inferred from just events, for speed. However, I think it's also worth mentioning that the reason all of this tate is thrown together in one giant "mutable state" object instead of a more traditional relational data model is that we need to support non-relational databases like Cassandra.

This particular distinction isn't very important, but I think these limitations explain a lot of our unintuitive design choices, so they're worth pointing out.